File: CoderOJ/ProbC.py
def min_delay(T, test_cases):
    res = []
    for _ in range(T):
        n, machines = test_cases[_]
        machines.sort(key=lambda x: x[0] + x[1])
        low, high = 1, 10**18
        while low < high:
            mid = (low + high) // 2
            if sum(max(0, (mid - m[0]) // m[1]) for m in machines) >= n:
                high = mid
            else:
                low = mid + 1
        res.append(high)
    return res


def take_input():
    T = int(input().strip())
    test_cases = []
    for _ in range(T):
        n = int(input().strip())
        machines = []
        for _ in range(5):
            a, b = map(int, input().strip().split())
            machines.append((a, b))
        test_cases.append((n, machines))
    return T, test_cases


T, test_cases = take_input()
print(min_delay(T, test_cases))

File: CoderOJ/ProbC.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<long long> min_delay(int T, vector<pair<int, vector<pair<int, int>>>> &test_cases)
{
    vector<long long> res;
    for (int i = 0; i < T; i++)
    {
        int n = test_cases[i].first;
        vector<pair<int, int>> &machines = test_cases[i].second;
        sort(machines.begin(), machines.end(), [](const pair<int, int> &a, const pair<int, int> &b)
             { return a.first + a.second < b.first + b.second; });

        long long low = 1, high = 1e18;
        while (low < high)
        {
            long long mid = (low + high) / 2;
            long long count = 0;
            for (const auto &m : machines)
            {
                count += max(0LL, (mid - m.first) / m.second);
            }
            if (count >= n)
            {
                high = mid;
            }
            else
            {
                low = mid + 1;
            }
        }
        res.push_back(high);
    }
    return res;
}

vector<pair<int, vector<pair<int, int>>>> take_input()
{
    int T;
    cin >> T;
    vector<pair<int, vector<pair<int, int>>>> test_cases;
    for (int i = 0; i < T; i++)
    {
        int n;
        cin >> n;
        vector<pair<int, int>> machines;
        for (int j = 0; j < 5; j++)
        {
            int a, b;
            cin >> a >> b;
            machines.push_back({a, b});
        }
        test_cases.push_back({n, machines});
    }
    return test_cases;
}

int main()
{
    vector<pair<int, vector<pair<int, int>>>> test_cases = take_input();
    vector<long long> result = min_delay(test_cases.size(), test_cases);
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << endl;
    }
    return 0;
}
